Replace manual parsing of head comments

SgtCoDFish opened the pull request:
This leaves more to the YAML parser so we have less chance of accidentally introducing a bug or security issue in our own logic.

Also adds preservation of footer comments as a bonus!

In addition, this fixes `klone init` and `klone add` which are currently broken on main when run in a directory with no `klone.yaml`. That bug was caused by the comment preservation logic!

inteon commented:
@SgtCoDFish
https://github.com/cert-manager/klone/pull/4 added the necessary tests that will make sure this PR is tested for some of the yaml weirdness that can happen.

SgtCoDFish committed:
replace manual parsing of head comments

This leaves more to the YAML parser so we have less chance of
accidentally introducing a bug or security issue in our own logic.

Also adds preservation of footer comments as a bonus!

Signed-off-by: Ashley Davis <ashley.davis@venafi.com>

SgtCoDFish commented:
It seems like the "weirdness" is that there's a difference between the following:

```yaml
# Comment

targets:
  target1:
    - ...
```

On the above, the comment is on the "document" and this PR will preserve it.

```yaml
# Comment
targets:
  target1:
    - ...
```

On this one, the coment is on the `targets` node and this PR (as initially written) doesn't preserve it. The comment is in `rawDocument.Content[0].Content[0].HeadComment` and not `rawDocument.HeadComment`

SgtCoDFish commented:
The latest commit passes the tests by adding a newline to the test cases so that the comments are now doc comments. I dunno about this... it makes sense to only preserve doc comments to me, and I can see why the yaml parser would treat the other comments differently. I'm not militant about it though

SgtCoDFish commented:
I think it'd be better to use a YAML parser for this but I don't think it's worth the effort coming up with a better solution here. Closing.

inteon commented:
> I think it'd be better to use a YAML parser for this but I don't think it's worth the effort coming up with a better solution here. Closing.

Thanks for testing and letting me know about the io.EOF error btw!
